tests/API/BasicCrudTest.php

```diff
@@ -2,58 +2,165 @@
-        return $client->getAll();
+    /**
+     * CRUD API client to test.
+     *
+     * @return mixed
+     */
+    abstract protected function getApiClient();
+
+    /**
+     * Data to use to create the test entity.
+     *
+     * @return array
+     */
+    abstract protected function getCreateBody();
+
+    /**
+     * Data to use to update the test entity.
+     *
+     * @return array
+     */
+    abstract protected function getUpdateBody();
+
+    /**
+     * Assertions for the created entity.
+     *
+     * @param array $created_entity - Created entity.
+     *
+     * @return mixed
+     */
+    abstract protected function afterCreate($created_entity);
+
+    /**
+     * Assertions for the updated entity.
+     *
+     * @param array $updated_entity - Updated entity.
+     *
+     * @return mixed
+     */
+    abstract protected function afterUpdate($updated_entity);
+
+    /**
+     * BasicCrudTest constructor.
+     * Sets up environment and domain value.
+     */
+    public function __construct()
+    {
+        parent::__construct();
+        $this->env = $this->getEnv();
+        $this->domain = $this->env['DOMAIN'];
+        $this->api = $this->getApiClient();
     }
 
-    protected function getId($entity) {
-        return $entity['id'];
+    /**
+     * Stub "get all entities" method.
+     * Can be overridden by child classes for specific test cases.
+     *
+     * @param array $created_entity - Created entity.
+     *
+     * @return mixed
+     */
+    protected function getAllEntities($created_entity)
+    {
+        return $this->api->getAll();
     }
 
-    public function testAll() {
-
-        $client = $this->getApiClient();
-
-        $options = $client->getApiClient()->get()->getGuzzleOptions();
+    /**
+     * Get the unique identifier for the entity.
+     *
+     * @param array $entity - Entity array.
+     *
+     * @return mixed
+     */
+    protected function getId($entity)
+    {
+        return $entity[$this->id_name];
+    }
 
+    /**
+     * Check that HTTP options have been set correctly.
+     */
+    public function testHttpOptions()
+    {
+        $options = $this->api->getApiClient()->get()->getGuzzleOptions();
         $this->assertArrayHasKey('base_uri', $options);
         $this->assertEquals("https://$this->domain/api/v2/", $options['base_uri']);
+    }
 
-        $created = $client->create($this->getCreateBody());
-
-        $all = $this->getAll($client, $created);
-
-        $found = false;
-        foreach ($all as $value) {
-            if ($this->getId($value) === $this->getId($created)) {
-                $found = true;
-                break;
+    /**
+     * All basic CRUD test assertions.
+     */
+    public function testAll()
+    {
+        // Test a generic "create entity" method for this API client.
+        $created_entity = $this->api->create($this->getCreateBody());
+        $created_entity_id = $this->getId($created_entity);
```

Comment:
since this line is used not in the line right next below this one, i'll swap this line and the one below.